src/opencl_autotune.c

```diff
@@ -59,6 +59,12 @@ size_t autotune_get_task_max_work_group_size(int use_local_memory,
 	else
 		max_available = get_device_max_lws(gpu_id);
 
+	if (cpu(device_info[gpu_id])) {
+#include <omp.h>
+		max_available = omp_get_max_threads();
+		get_power_of_two(max_available);
+	}
```

Comment:
I guess we should try to research this. I'll google around a little.

Reply:
We can add a multiplication, such as num_threads x 2 or num_threads x 4. However, nothing beyond this makes sense.

Reply:
Super has 32 cores, and auto-tunes LWS to anywhere between 128..1024 with NT-opencl, ending up in more or less the same performance. It doesn't seem to want lower than 128 but if I peg it to 16 (so equal to only half the number of cores), it performs as good anyway.
Lower than 4 is really bad, 4 is OK and >=8 is full performance.

So no answers from that really.

Reply:
My old 4-core/8-thread machine tuned to LWS=8 for wpapsk and 16 for nt. Those numbers are "logical" to me - one OpenCL thread per CPU thread, or two (somewhat like `OMP_SCALE 2`) for a fast format.

Reply:
> if I peg it to 16 (so equal to only half the number of cores), it performs as good anyway.

Perhaps that actually ends up as two groups of 16 running concurrently on 32 cores? Precise grouping is mostly about memory access and barriers, much more delicate on a GPU.

Reply:
IMO: a multiplier (perhaps 4) combined with the fact that the autotune is still doing its job (up to the new max) should be enough to ensure that the patch is correct.

Reply:
x4 sounds about right from what we know right now. But I'll check a couple machine out on Monday.

Reply:
Also, if `get_max_compute_units()` returns anything _higher than 1_ (super's runtime stupidly say 1 even for 32 cores), that figure should be used and this OpenMP check can be skipped. But if we get 1 from that, the OpenMP check should be tried instead. It has other flaws, such as OpenMP not always available, or `OMP_NUM_THREADS` limited.

BTW the omp.h header include should be moved up to where it belongs, and must be guarded with `#ifdef _OPENMP` like elsewhere. Then this whole new (`else`) block of code should also be inside `#ifdef _OPENMP`.

If all else fails, we should set it to something sensible, I guess 16 at the very least.

Reply:
Also, if `get_max_compute_units()` returns anything _higher than 1_ (super's runtime stupidly say 1 even for 32 cores), that figure should be used and this OpenMP check can be skipped. But if we get 1 from that, the OpenMP check should be tried instead. It has other flaws, such as OpenMP not always available, or `OMP_NUM_THREADS` limited.

BTW the omp.h header include should be moved up to where it belongs, and must be guarded with `#ifdef _OPENMP` like elsewhere. Then this whole new (`else`) block of code should also be inside `#ifdef _OPENMP`.

If all else fails, we should set it to something sensible, I guess 16 at the very least.